clang/lib/CIR/Dialect/IR/CIRAttrs.cpp

```diff
@@ -498,6 +499,63 @@ LogicalResult DynamicCastInfoAttr::verify(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// OpenCLKernelMetadataAttr definitions
+//===----------------------------------------------------------------------===//
+
+LogicalResult OpenCLKernelMetadataAttr::verify(
+    ::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
+    ArrayAttr workGroupSizeHint, ArrayAttr reqdWorkGroupSize,
+    TypeAttr vecTypeHint, std::optional<bool> vecTypeHintSignedness,
+    IntegerAttr intelReqdSubGroupSize) {
+  // If no field is present, the attribute is considered invalid.
+  if (!workGroupSizeHint && !reqdWorkGroupSize && !vecTypeHint &&
+      !vecTypeHintSignedness && !intelReqdSubGroupSize) {
+    emitError() << "metadata attribute without any field present is invalid";
+    return failure();
+  }
+
+  // Check for 3-dim integer tuples
+  auto is3dimIntTuple = [](ArrayAttr arr) {
+    auto isInt = [](Attribute dim) { return mlir::isa<IntegerAttr>(dim); };
+    return arr.size() == 3 && llvm::all_of(arr, isInt);
+  };
+  if (workGroupSizeHint && !is3dimIntTuple(workGroupSizeHint)) {
+    emitError() << "work_group_size_hint must have exactly 3 integer elements";
+    return failure();
+  }
+  if (reqdWorkGroupSize && !is3dimIntTuple(reqdWorkGroupSize)) {
+    emitError() << "reqd_work_group_size must have exactly 3 integer elements";
+    return failure();
+  }
+
+  // Check for co-presence of vecTypeHintSignedness
+  if (!!vecTypeHint != vecTypeHintSignedness.has_value()) {
+    emitError() << "vec_type_hint_signedness should be present if and only if "
+                   "vec_type_hint is set";
+    return failure();
+  }
+
+  if (vecTypeHint) {
+    Type vecTypeHintValue = vecTypeHint.getValue();
+    if (mlir::isa<cir::CIRDialect>(vecTypeHintValue.getDialect())) {
+      // Check for signedness alignment in CIR
+      if (isSignedHint(vecTypeHintValue) != vecTypeHintSignedness) {
+        emitError() << "vec_type_hint_signedness must match the signedness of "
+                       "the vec_type_hint type";
+        return failure();
+      }
+      // Check for the dialect of type hint
+    } else if (!LLVM::isCompatibleType(vecTypeHintValue)) {
+      emitError() << "vec_type_hint must be a type from the CIR or LLVM "
+                     "dialect";
```

Comment:
Not checked in `invalid.cir`.

Reply:
This one is a bit difficult to test as we don't load other dialects in the test case.

Reply:
You could perhaps create an `invalid-opencl-metadata.cir` and load all necessary stuff?